Twin IP display, battery percentage

MoffKalast opened the pull request:
So it does seem like setting the static eth0 IP messses up deteciton because it's currently set up as "if eth0 ip is > 7 chars then use that instead of wlan" for some reason.

In my case here, wlan0 is 192.168.1.11 on my local network and eth0 is set up as the lidar IP.

Dynamic eth0 (working as intended):
![20220303_173949](https://user-images.githubusercontent.com/9977799/156629738-1e2ac580-9b1d-47b0-8189-d931bdedda2e.jpg)

Static eth0 (showing eth0 instead of wlan):
![20220303_174430](https://user-images.githubusercontent.com/9977799/156629734-e74736fe-31c6-484d-b54f-d6e40dd7d074.jpg)

My suggestion is "porque no los dos?"
![20220303_193011](https://user-images.githubusercontent.com/9977799/156629743-95a511b2-1017-4abd-8f47-a768219adc25.jpg)

We seem to have 8 effective lines to write on and currently only use 6, so I've expanded it a bit.

I've added a battery percentage display in this PR since the voltage-only display is not that useful to an untrained person. The charging voltage is now set to 27.8 since it was set to rather low, with CHRG flashing slower and less often than LOW.

MoffKalast commented:
Test with eth0 disconnected:
![20220303_194341](https://user-images.githubusercontent.com/9977799/156631611-c9da41b6-96a2-48c5-9680-ba2e6116700a.jpg)

Test with eth0 on local DHCP:
![20220303_194501](https://user-images.githubusercontent.com/9977799/156631606-564a82c9-c29a-44e2-81e2-513aacecb6a8.jpg)

mjstn commented:
I had made a brief attempt to get the logic right but there are many corner cases. Thanks Vid.
My top goal was to fix the backwards printing of info due to incorrect display detection and to add motor power on off.
My attempt to fix IP detection was done 'along the way' but I even felt it may not hold although it worked for my needs.
I am testing and reviewing this change now.

mjstn commented:
The > 7 chars hack was that since we were only showing ONE IP I decided that if there was an IP on eth0 that would be what we show.  Of course that breaks when some fool starts to use a network lidar ... lol.    I like the showing of BOTH although some care may be required if somebody puts a switch on the Pi ethernet and messes with network stuff but that is out of scope and if such is done that is his issue to sort out the display, NOT ours.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -55,17 +55,18 @@ int main(int argc, char** argv) {
 
 // Define a level where we think the charger is plugged in at this time
-#define  BAT_CHARGING_LEVEL  27.0                // Indicate battery is on charger for high voltages
+#define  BAT_CHARGING_LEVEL  27.8                // Indicate battery is on charger for high voltages
```
I have noticed this.  It needs to be above 27, just how much is the question.   I would have picked 12.5 but basically there is no 'best' answer so 27.8 is ok here.

MoffKalast replied:
Assuming we even want to keep this, since it's a bit of a hack. I later realized the image I was testing on had an outdated ubiquity_motor, so it was only reaching 100% charge at 28V which is of course ludicrous. I'll need to re-test with the new master and see what the threshold actually is.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -864,32 +871,35 @@ int main(int argc, char **argv)
+        stream << " CHRG";
+      }
+      else{
+        stream << "V" << g_batteryPercentage;
```
This was my main concern and this looks great.  We really must show  LOW and CHRG and in between showing the percentage is a nice improvement in place of  just  'OK'.  Like that we say LOW and as we 'blink' that it alternates from percentage to LOW.   Noice    I will pull and try this myself now on noetic with lidar.  I use pifi and thus my wifi IP will be my house lan.

mjstn reviewed (approved):
This looks like a nice improvement.  I have not tested all 4 cases but they do need to be tested.
The missing cases I have not tried yet are these:

- Wifi AP mode shows proper IP for top line
- What does second line show when network is setup for the ethernet Lidar

I do not think we should suppress eth0 when a lidar and have not tried that.    Somebody may put in a router there so they can have a hardline to the Magni at the expense of being forced to use the IP we use on that network.   If eth0 has an IP we should show it.

So this is indeed an interesting tradeoff.  If we show the IP for eth0 when it is not possible nor does a user even care about that internal only IP is that a point of confusion?

Anyway, leave it to always show an eth0 IP and lets not try to outguess everyone just yet.

anfederman commented:
Idea why not a  4 letter display  on the last line showing what hardware is
there:

W/I. For WAP or Internet
C   camera
S Sonar
L  lidar

On Sun, Mar 6, 2022, 1:23 PM Mark Johnston ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> This looks like a nice improvement. I have not tested all 4 cases but they
> do need to be tested.
> The missing cases I have not tried yet are these:
>
>    - Wifi AP mode shows proper IP for top line
>    - What does second line show when network is setup for the ethernet
>    Lidar
>
> I do not think we should suppress eth0 when a lidar and have not tried
> that. Somebody may put in a router there so they can have a hardline to the
> Magni at the expense of being forced to use the IP we use on that network.
> If eth0 has an IP we should show it.
>
> So this is indeed an interesting tradeoff. If we show the IP for eth0 when
> it is not possible nor does a user even care about that internal only IP is
> that a point of confusion?
>
> Anyway, leave it to always show an eth0 IP and lets not try to outguess
> everyone just yet.
> ------------------------------
>
> In src/oled_display_node.cpp
> <https://github.com/UbiquityRobotics/oled_display_node/pull/22#discussion_r820285744>
> :
>
> > -        std::string battText = "Bat " + stream.str();
> -        dispError |= dispOled_writeText(&g_oledDisplayCtx, DISP_LINE_BATT_VOLTS, 1, DISP_TEXT_START_MODE, battText.c_str());
> -        ros::Duration(updateDelay).sleep();
> +      // Control what shows up after the voltage reading.
> +      // If you change the text, use same number of chars so display does not move around on the line
> +      std::stringstream stream;
> +      stream << std::fixed << std::setprecision(1) << g_batteryVoltage;
> +
> +      if (g_batteryVoltage <= BAT_LOW_LEVEL && (loopIdx % 2) == 0){
> +        stream << "V LOW";
> +      }
> +      else if (g_batteryVoltage >= BAT_CHARGING_LEVEL && (loopIdx % 3) == 0){
> +        stream << " CHRG";
> +      }
> +      else{
> +        stream << "V" << g_batteryPercentage;
>
> This was my main concern and this looks great. We really must show LOW and
> CHRG and in between showing the percentage is a nice improvement in place
> of just 'OK'. Like that we say LOW and as we 'blink' that it alternates
> from percentage to LOW. Noice I will pull and try this myself now on noetic
> with lidar. I use pifi and thus my wifi IP will be my house lan.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UbiquityRobotics/oled_display_node/pull/22#pullrequestreview-901118682>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHZ376NZHBGO74OFC6DJY3U6UO4JANCNFSM5P3JMH5A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -864,32 +871,35 @@ int main(int argc, char **argv)
+      if (g_batteryVoltage <= BAT_LOW_LEVEL && (loopIdx % 2) == 0){
+        stream << "V LOW";
+      }
+      else if (g_batteryVoltage >= BAT_CHARGING_LEVEL && (loopIdx % 3) == 0){
```
PROBLEM HERE  I just ran on charger for yet another case.   Don't show the percentage when in CHRG display.  It is of near zero value and as it stands right now you get this packed display that is all run together.  It ends up when 100% is showing having   27.8V100%  so looks very bad.
The purpose of 'blinking' when LOW was to attract attention.   When voltage is in 'Charging' range just show CHRG and don't show percent.  My experience with these batteries of assorted ages is the voltage it sits at while charging is of little value it can read 29.2V for super weak battery but not at all be done charging.     If you really feel percentage is important to you when at high V charging then you will need to fix the 'V' that leads to all letters run together for 100% display.   
Also the  V should be gotten rid of anyway.  Our display is so cramped for space  saying V is not useful because the V is implied and our doc on this should say that value is volts.

mjstn commented:
This is specifically what I am asking for.  If not this then make some change to get rid of the V when percent is 100

      // Control what shows up after the voltage reading.
      // If you change the text, use same number of chars so display does not move around on the line
      std::stringstream stream;
      stream << std::fixed << std::setprecision(1) << g_batteryVoltage;

      if (g_batteryVoltage <= BAT_LOW_LEVEL && (loopIdx % 2) == 0){
        stream << "V LOW";
      }
      else if (g_batteryVoltage >= BAT_CHARGING_LEVEL){           <<< **This is the only actual change**
        stream << " CHRG";
      }
      else{
        stream << "V" << g_batteryPercentage;
      }

MoffKalast commented:
> Anyway, leave it to always show an eth0 IP and lets not try to outguess everyone just yet.

Just to be clear, the line isn't suppressed in this node, it's just not shown when the adapter has no actual IP attached to it. E.g. when a cable is disconnected and there's no static IP. We could show something like a "NO ETH0 IP" line instead to make it clearer I suppose.

> if somebody puts a switch on the Pi ethernet and messes with network stuff

No that should be fine, we show the adapter address whatever it may be, it doesn't get any simpler than that. The difference may be what the roscore master is actually bound to but that's for the user to know if they're on a custom setup.


> 4 letter display  on the last line showing what hardware is there

Yeah I was thinking of something along those lines, since we could put more info there than just motors. We'd need to have good ways of figuring out those data points though, which may no be straightforward.

Sure we could say subscribe to /sonars, /scan, /raspicam_node and display them as missing if a message isn't received every X seconds, but those can be remapped and may put more strain on the messaging system than needed since they're all high data rate topics. A simpler way would be to check if any topic of a type of LaserScan, Sonar, or CompressedImage with a hz > 0 exists , but that's not super reliable either imo.

> If not this then make some change to get rid of the V when percent is 100

Reading the IP lines, I've noticed that we seem to have 14 letters per line and not 13 as I thought, so it may be possible to fit another space in there without changing anything else.

MoffKalast commented:
Hmm, so it seems that the [percentage fix in ubiquity_motor](https://github.com/UbiquityRobotics/ubiquity_motor/tree/battery-fix) has only been merged into kinetic-devel and not noetic-devel. We'll need to add that before this can be merged, since right now the percentages are completely wrong.

MoffKalast commented:
Alright so I've fixed the battery spacing and I've also edited the last line to display the firmware version along with the motor power. This also ties into https://github.com/UbiquityRobotics/ubiquity_motor/pull/168 which adds the firmware publishing to a String topic, but in case that's not available it should just display `FW:---`

Also fixed some indents and set them to 4 spaces, just to clean things up a bit.

![image](https://user-images.githubusercontent.com/9977799/158851749-0bfd7a66-9583-4e95-8300-426e69fd3861.png)

I've also just now noticed that Bat is title case and MOT is all caps, so I guess we should go with either BAT+MOT or Bat+Mot.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -775,148 +773,162 @@ void displayApiCallback(const oled_display_node::DisplayOutput::ConstPtr& msg)
+
+        // If you change the text, use same number of chars so display does not move around on the line
+        std::stringstream infoStream;
+        infoStream << "Mot:" << g_motorPowerActive << " FW:" << g_firmwareVersion;
```
Ok.  I guess --- is ok for indication of no callback.   This will in fact happen to people who enable the display and pull new code so this will not just be for leaches who the oled driver.   I silenced some parts of the display s that blank was better than stuff but the argument should be we do what is forward looking for the product so it's ok.

MoffKalast replied:
Yeah we could easily do blank I suppose, but this seemed more indicative at the time.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -775,148 +773,162 @@ void displayApiCallback(const oled_display_node::DisplayOutput::ConstPtr& msg)
+            if (g_batteryVoltage <= BAT_LOW_LEVEL && (loopIdx % 2) == 0){
+                stream << "V  LOW";
+            }
+            else if (g_batteryVoltage >= BAT_CHARGING_LEVEL && (loopIdx % 3) != 0){
```
This looks nice now.

MoffKalast replied:
Yeah I've switched it to sometimes flash the percentage and usually show CHRG instead of the opposite, since it's probably more in line with what's intended to be shown.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -55,9 +55,10 @@ int main(int argc, char** argv) {
+#define  DISP_LINE_IP_WLAN     1
+#define  DISP_LINE_IP_ETH     3
+#define  DISP_LINE_BATT_VOLTS  5
+#define  DISP_LINE_MOTOR_POWER 7 
```
I am glad you did not remove the forcing of one display type, that would have backed out an important change.

MoffKalast replied:
Yeah I didn't really go into that, if there are any changes to it it's because of the indentation fix I ran.

mjstn reviewed (approved):
Nice work this is getting really busy but also really useful!

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -775,148 +773,162 @@ void displayApiCallback(const oled_display_node::DisplayOutput::ConstPtr& msg)
  */
 void batteryStateApiCallback(const sensor_msgs::BatteryState::ConstPtr& msg)
 {
+    std::string percent = std::to_string((int)((msg->percentage * 100) + 0.5))+"%";
```
I was trying this on kinetic image and sadly got make error.   We must be sure this runs on Kinetic

/home/ubuntu/catkin_ws/src/oled_display_node/src/oled_display_node.cpp:776:27: error: 'to_string' is not a member of 'std'
     std::string percent = std::to_string((int)((msg->percentage * 100) + 0.5))+"%";

MoffKalast replied:
Hmm that's odd, I've had some similar issues during dev on noetic as well, I'll check what happens on kinetic.

MoffKalast replied:
Huh, apparently it may be a compiler bug with older gccs. The workaround is to use the stringstreams which seems kind of weird but okay.

mjstn replied:
I too hate dorking with the stringstreams but sadly, that is more 'acceptable' although awkward.

MoffKalast commented:
@mjstn I've added a patch for kinetic.

anfederman commented:
Is this OLED code ready for testing on the 3-19 image?

On Sat, Mar 19, 2022, 2:53 PM Mark Johnston ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/oled_display_node.cpp
> <https://github.com/UbiquityRobotics/oled_display_node/pull/22#discussion_r830531250>
> :
>
> > @@ -775,148 +773,162 @@ void displayApiCallback(const oled_display_node::DisplayOutput::ConstPtr& msg)
>   */
>  void batteryStateApiCallback(const sensor_msgs::BatteryState::ConstPtr& msg)
>  {
> +    std::string percent = std::to_string((int)((msg->percentage * 100) + 0.5))+"%";
>
> I too hate dorking with the stringstreams but sadly, that is more
> 'acceptable' although awkward.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UbiquityRobotics/oled_display_node/pull/22#discussion_r830531250>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHZ372BKYQ6W6Z74Y62ZFDVAZEHPANCNFSM5P3JMH5A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

mjstn commented:
Alan:  No this code is not in March 19 noetic image.   This code 'missed the boat' and will be for future images and hopefully apt upgrade once this is in master.

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -136,8 +137,10 @@ int main(int argc, char** argv) {
 
 // Some limited state for display
+std::string g_batteryPercentage = "    ";
+std::string g_firmwareVersion = "---";
```
For Kinetic images we are likely going to have to do an entirely different ubiquity_motor PR to pull in what you did for noetic ubiquity_motor.  To me it seems acceptable that firmware rev is an undone feature at this time.

MoffKalast replied:
Yeah we could add that to kinetic too, though given that we'll be shipping only noetic in a month or two I think it's not super problematic. Plus we'll be able to identify people still using kinetic by the fw number not showing 😄

mjstn commented on src/oled_display_node.cpp:
```diff
@@ -162,53 +165,48 @@ extern  int dispOled_writeText(dispCtx_t *dispCtx, uint8_t line, uint8_t segment
-  ipAddress.assign(mainIpAddress);
-  return;
+// getIpAddressses()
+// We show both the WLAN0 and ETH0 adapter IPs since both can be required by the user
```
I will prey to the gods for this iteration.   My own tests look rather good so far but I bet there is SOME combination that still is not happy.  Still, this is a nice improvement

MoffKalast replied:
Well we might need to try and see what happens if we attach a few usb to ethernet adapters and wifi dongles ha.

mjstn reviewed (approved):
This looks great on Noetic as long as ubiquity_motor is also pulled so that firmware revision is available.    I think since this is a cross-repo issue that this is an ok to do it in the OLED node. Non-ideal but ok.   Best case would be suppression of showing  FW:--- if we never have gotten the callback for firmware but that is a nitpick and I leave that up to you.  I say ok as it stands.
